File: kotlin/src/dk/leetcode/MaximumDifferenceBetweenNodeAndAncestor.kt
package dk.leetcode
// https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/

class MaximumDifferenceBetweenNodeAndAncestor {
    protected class MM(val mn: Int, val mx: Int) {
        inline fun combine(v: Int) = MM(minOf(mn, v), maxOf(mx, v))
        inline fun max(v: Int) = arrayOf(v - mn, mx - v, mx - v, v - mn).max()!!
    }

    protected fun fndMaxDiff(top: MM, root: TreeNode?): Int {
        if (root != null) {
            val s = top.max(root.`val`)
            val ntop = top.combine(root.`val`)
            val l = fndMaxDiff(ntop, root.left)
            val r = fndMaxDiff(ntop, root.right)
            return maxOf(s, l, r)
        } else {
            return 0
        }
    }

    fun maxAncestorDiff(root: TreeNode?) = fndMaxDiff(MM(root!!.`val`, root!!.`val`), root)
}

File: c/lc/maximum-difference-between-node-and-ancestor.cc
// https://leetcode.com/problems/maximum-difference-between-node-and-ancestor
class Solution {
    struct Rng {
        int mn, mx;
        Rng(int v): mn{v}, mx{v} {
        }
        Rng(const Rng &o, int v): mn{o.mn}, mx{o.mx} {
            mn = min(mn, v);
            mx = max(mx, v);
        }
        int dist(int v) const {
            return max(abs(mn - v), abs(mx - v));
        }
    };
    int diff(TreeNode* root, const Rng &rng) {
        if (root) {
            int self = rng.dist(root->val);
            const Rng nrng(rng, root->val);
            int left = diff(root->left, nrng);
            int right = diff(root->right, nrng);
            return max({self, left, right});
        } else {
            return 0;
        }
    }
public:
    int maxAncestorDiff(TreeNode* root) {
        const Rng rng(root->val);
        return diff(root, rng);
    }
};
